File: team-ai/app/transport/grpc/servicers/ai.py
"""AIService gRPC servicer — provides ShoppingAssistant, MagicListing, and ChatCopilot.

Adheres to the platform-core gRPC contract (platform.ai.v1.AIService).
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Callable

import grpc
from loguru import logger

from app.modules.business.ai_assistant.schemas import (
    ChatCopilotRequest,
    MagicListingRequest,
    ReviewItem,
    ShoppingAssistantRequest,
    SummarizeReviewsRequest,
)
from app.modules.business.ai_assistant.service import AIAssistantService
from app.transport.grpc._pb.platform.ai.v1 import ai_pb2, ai_pb2_grpc

if TYPE_CHECKING:
    pass

AIProvider = Callable[[], AIAssistantService]


class AIServicer(ai_pb2_grpc.AIServiceServicer):
    def __init__(self, ai_provider: AIProvider | None = None) -> None:
        self._ai_provider = ai_provider or (lambda: AIAssistantService())

    async def ShoppingAssistant(
        self,
        request: ai_pb2.ShoppingAssistantRequest,
        context: grpc.aio.ServicerContext,
    ) -> ai_pb2.ShoppingAssistantResponse:
        try:
            ai_service = self._ai_provider()
            req = ShoppingAssistantRequest(
                message=request.message,
                user_id=request.user_id,
                previous_context=list(request.previous_context),
            )
            result = await ai_service.shopping_assistant(req)

            product_cards = [
                ai_pb2.ProductCardSnippet(
                    listing_id=card.listing_id,
                    title=card.title,
                    price=card.price,
                    currency=card.currency,
                    image_url=card.image_url,
                    discount_rate=card.discount_rate,
                    rating_text=card.rating_text,
                )
                for card in result.product_cards
            ]

            return ai_pb2.ShoppingAssistantResponse(
                reply_text=result.reply_text,
                product_cards=product_cards,
                suggested_followups=result.suggested_followups,
            )
        except Exception as exc:
            logger.exception("grpc.ShoppingAssistant.error error={}", exc)
            await context.abort(grpc.StatusCode.INTERNAL, str(exc))
            raise

    async def MagicListing(
        self,
        request: ai_pb2.MagicListingRequest,
        context: grpc.aio.ServicerContext,
    ) -> ai_pb2.MagicListingResponse:
        try:
            ai_service = self._ai_provider()
            req = MagicListingRequest(
                title_hint=request.title_hint,
                category_hint=request.category_hint,
                image_url=request.image_url,
            )
            result = await ai_service.magic_listing(req)

            return ai_pb2.MagicListingResponse(
                generated_title=result.generated_title,
                generated_description=result.generated_description,
                suggested_category_id=result.suggested_category_id,
                suggested_price_min=result.suggested_price_min,
                suggested_price_max=result.suggested_price_max,
                highlight_tags=result.highlight_tags,
            )
        except Exception as exc:
            logger.exception("grpc.MagicListing.error error={}", exc)
            await context.abort(grpc.StatusCode.INTERNAL, str(exc))
            raise

    async def ChatCopilot(
        self,
        request: ai_pb2.ChatCopilotRequest,
        context: grpc.aio.ServicerContext,
    ) -> ai_pb2.ChatCopilotResponse:
        try:
            ai_service = self._ai_provider()
            req = ChatCopilotRequest(
                seller_id=request.seller_id,
                buyer_message=request.buyer_message,
                listing_id=request.listing_id,
            )
            result = await ai_service.chat_copilot(req)

            return ai_pb2.ChatCopilotResponse(
                quick_replies=result.quick_replies,
            )
        except Exception as exc:
            logger.exception("grpc.ChatCopilot.error error={}", exc)
            await context.abort(grpc.StatusCode.INTERNAL, str(exc))
            raise

    async def SummarizeReviews(
        self,
        request: ai_pb2.SummarizeReviewsRequest,
        context: grpc.aio.ServicerContext,
    ) -> ai_pb2.SummarizeReviewsResponse:
        try:
            ai_service = self._ai_provider()
            req = SummarizeReviewsRequest(
                listing_id=request.listing_id,
                reviews=[
                    ReviewItem(rating=r.rating, comment=r.comment)
                    for r in request.reviews
                ],
            )
            result = await ai_service.summarize_reviews(req)

            return ai_pb2.SummarizeReviewsResponse(
                summary=result.summary,
                pros=result.pros,
                cons=result.cons,
                sentiment=result.sentiment,
            )
        except Exception as exc:
            logger.exception("grpc.SummarizeReviews.error error={}", exc)
            await context.abort(grpc.StatusCode.INTERNAL, str(exc))
            raise

File: team-frontend/src/lib/gateway/ai.ts
/**
 * Server-side AI data functions. They wrap the gateway AIService RPCs and map
 * proto messages to plain view models (bigint → number) so React components and
 * Server Actions stay free of proto types. The frontend never calls team-ai
 * directly — every call goes through the gateway (Rule 1).
 */
import "server-only";

import { makeClients } from "./client.js";
import { getToken } from "./session.js";

// gateway builds request-scoped clients carrying the caller's session token.
function gateway() {
  return makeClients(getToken());
}

export interface ViewProductCard {
  listingId: string;
  title: string;
  price: number;
  currency: string;
  imageUrl: string;
  discountRate: number;
  ratingText: string;
}

export interface ViewAssistantReply {
  replyText: string;
  productCards: ViewProductCard[];
  suggestedFollowups: string[];
}

export interface ViewMagicListing {
  generatedTitle: string;
  generatedDescription: string;
  suggestedCategoryId: string;
  suggestedPriceMin: number;
  suggestedPriceMax: number;
  highlightTags: string[];
}

/** Shopping Assistant (RAG chatbot): question → reply + product cards + followups. */
export async function shoppingAssistant(
  message: string,
  previousContext: string[] = [],
  userId = "",
): Promise<ViewAssistantReply> {
  const res = await gateway().ai.shoppingAssistant({
    userId,
    message,
    previousContext,
  });
  return {
    replyText: res.replyText,
    productCards: res.productCards.map((c) => ({
      listingId: c.listingId,
      title: c.title,
      price: Number(c.price),
      currency: c.currency,
      imageUrl: c.imageUrl,
      discountRate: c.discountRate,
      ratingText: c.ratingText,
    })),
    suggestedFollowups: res.suggestedFollowups,
  };
}

/** Magic Listing ("tạo tin nhanh"): title hint → SEO title/description/category/price. */
export async function magicListing(
  titleHint: string,
  categoryHint = "",
  imageUrl = "",
): Promise<ViewMagicListing> {
  const res = await gateway().ai.magicListing({
    titleHint,
    categoryHint,
    imageUrl,
  });
  return {
    generatedTitle: res.generatedTitle,
    generatedDescription: res.generatedDescription,
    suggestedCategoryId: res.suggestedCategoryId,
    suggestedPriceMin: Number(res.suggestedPriceMin),
    suggestedPriceMax: Number(res.suggestedPriceMax),
    highlightTags: res.highlightTags,
  };
}

/** Chat Copilot: buyer message → up to 3 quick seller replies. */
export async function chatCopilot(
  sellerId: string,
  buyerMessage: string,
  listingId = "",
): Promise<string[]> {
  const res = await gateway().ai.chatCopilot({
    sellerId,
    buyerMessage,
    listingId,
  });
  return res.quickReplies;
}

export interface ViewReviewSummary {
  summary: string;
  pros: string[];
  cons: string[];
  sentiment: string;
}

/** AI review summary: condense a listing's reviews into pros/cons + sentiment. */
export async function summarizeReviews(
  listingId: string,
  reviews: { rating: number; comment: string }[],
): Promise<ViewReviewSummary | null> {
  if (reviews.length === 0) return null;
  try {
    const res = await gateway().ai.summarizeReviews({
      listingId,
      reviews: reviews.map((r) => ({ rating: r.rating, comment: r.comment })),
    });
    return {
      summary: res.summary,
      pros: res.pros,
      cons: res.cons,
      sentiment: res.sentiment,
    };
  } catch {
    return null; // team-ai unavailable — hide the block
  }
}
